src/gc/windows/gcenv.windows.cpp

```diff
@@ -30,6 +30,191 @@ typedef BOOL (WINAPI *PQUERY_INFORMATION_JOB_OBJECT)(HANDLE jobHandle, JOBOBJECT
+    bBuffer = new (std::nothrow) BYTE[ cbSLPIEx ];
+    if (bBuffer == NULL)
+        return false;
+
+    pSLPIEx = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *)bBuffer;
+    if (!GetLogicalProcessorInformationEx(RelationGroup, pSLPIEx, &cbSLPIEx))
+    {
+        delete[] bBuffer;
+        return false;
+    }
+
+    pRecord = pSLPIEx;
+    while (byteOffset < cbSLPIEx)
+    {
+        if (pRecord->Relationship == RelationGroup)
+        {
+            g_nGroups = pRecord->Group.ActiveGroupCount;
+            break;
+        }
+        byteOffset += pRecord->Size;
+        pRecord = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *)(bBuffer + byteOffset);
+    }
+
+    g_CPUGroupInfoArray = new (std::nothrow) CPU_Group_Info[g_nGroups];
+    if (g_CPUGroupInfoArray == NULL) 
+    {
+        delete[] bBuffer;
+        return false;
+    }
+
+    for (DWORD i = 0; i < g_nGroups; i++)
+    {
+        g_CPUGroupInfoArray[i].nr_active   = (WORD)pRecord->Group.GroupInfo[i].ActiveProcessorCount;
+        g_CPUGroupInfoArray[i].active_mask = pRecord->Group.GroupInfo[i].ActiveProcessorMask;
+        g_nProcessors += g_CPUGroupInfoArray[i].nr_active;
+        dwWeight = LCM(dwWeight, (DWORD)g_CPUGroupInfoArray[i].nr_active);
+    }
+
+    // The number of threads per group that can be supported will depend on the number of CPU groups
+    // and the number of LPs within each processor group. For example, when the number of LPs in
+    // CPU groups is the same and is 64, the number of threads per group before weight overflow
+    // would be 2^32/2^6 = 2^26 (64M threads)
+    for (DWORD i = 0; i < g_nGroups; i++)
+    {
+        g_CPUGroupInfoArray[i].groupWeight = dwWeight / (DWORD)g_CPUGroupInfoArray[i].nr_active;
+        g_CPUGroupInfoArray[i].activeThreadWeight = 0;
+    }
+
+    delete[] bBuffer;  // done with it; free it
+    return true;
+#else
+    return false;
+#endif
+}
+
+bool InitCPUGroupInfoRange()
+{
+#if (defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_))
+    WORD begin   = 0;
+    WORD nr_proc = 0;
+
+    for (WORD i = 0; i < g_nGroups; i++) 
+    {
+        nr_proc += g_CPUGroupInfoArray[i].nr_active;
+        g_CPUGroupInfoArray[i].begin = begin;
+        g_CPUGroupInfoArray[i].end   = nr_proc - 1;
+        begin = nr_proc;
+    }
+
+    return true;
+#else
+    return false;
+#endif
+}
+
+void InitCPUGroupInfo()
+{
+    g_fEnableGCCPUGroups = false;
+
+#if (defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_))
+    if (!g_fIsCPUGroupEnabledByConfig)
+        return;
+
+    if (!InitCPUGroupInfoArray())
+        return;
+
+    if (!InitCPUGroupInfoRange())
+        return;
+
+    // only enable CPU groups if more than one group exists
+    g_fEnableGCCPUGroups = g_nGroups > 1;
+#endif // _TARGET_AMD64_ || _TARGET_ARM64_
+
+    // Determine if the process is affinitized to a single processor (or if the system has a single processor)
+    DWORD_PTR processAffinityMask, systemAffinityMask;
+    if (GetProcessAffinityMask(
+#ifdef _WIN32
```

Comment:
This should not be need. `_WIN32` is always defined on Windows, and this is Windows-specific file.